data/transactions/logic/eval.go

```diff
@@ -1543,14 +1580,37 @@ func (cx *evalContext) txnFieldToStack(txn *transactions.Transaction, field TxnF
 		sv.Bytes = txn.FreezeAccount[:]
 	case FreezeAssetFrozen:
 		sv.Uint = boolToUint(txn.AssetFrozen)
+	case Scratch:
+		if txn.Type != protocol.ApplicationCallTx {
+			err = fmt.Errorf("can't use Scratch txn field on non-app call txn with index %d", groupIndex)
+			return
+		} else if cx.Txn.Lsig.Logic != nil {
+			err = fmt.Errorf("can't use Scratch txn field from within a LogicSig")
+			return
+		} else if arrayFieldIdx >= 256 {
+			err = fmt.Errorf("invalid Scratch index %d", arrayFieldIdx)
+			return
+		} else if groupIndex == cx.GroupIndex {
+			err = fmt.Errorf("can't use Scratch txn field on self, use load instead")
+			return
+		} else if groupIndex > cx.GroupIndex {
+			err = fmt.Errorf("can't get future Scratch from txn with index %d", groupIndex)
+			return
+		}
+		val := cx.PastSideEffects[groupIndex].GetScratchValue(uint8(arrayFieldIdx))
+		if val.Bytes != nil {
+			sv.Bytes = val.Bytes
+		} else {
+			sv.Uint = val.Uint
+		}
```

Comment:
Perhaps this means that instead of a clone() func on StackValue, you should add an assign(StackValue).  It would not allocate a StackValue to return it, it would assign the given StackValue into the receiver, copying the bytes field.

(Is GetScratchValue currently using clone()?)

Reply:
Yes! GetScratchValue is using clone().

Reply:
Hmm not sure why I didn't just assign to `sv` here. I'll simplify this to `sv = cx.PastSideEffects[...]`.